R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+#' This function creates a plot of the variables over time and returns a
+#' vectors of the incidence numbers for each age group.
+#'
+#' @param object An object of the class age_model.
+#' @param times (vector) time sequence over which to solve the model.
+#'        Must be of the form seq(t_start,t_end,by=t_step).
+#' @param is_plot (logical) whether plots for the compartments for the
+#'        different age groups are drawn.
+#'
+#' @return data frame of time and outputs with incidence numbers for each
+#'         age group.
+setMethod(
+  'simulate', 'age_model',
+  function(object, times, is_plot) {
+    if(!is.double(times)){
+      stop('Evaluation times of the model storage format must be a vector.')
+    }
+    if(!is.logical(is_plot)){
+      stop('Trigger for plotting must be logical.')
+    }
+    age <- object@n_age_categories
+
+    #set initial state vector
+    state <- c(S = get_parameters(object)$S0,
+               E = get_parameters(object)$E0,
+               I = get_parameters(object)$I0,
+               R = get_parameters(object)$R0)
+
+    #set parameters vector
+    parameters <- c(a = get_parameters(object)$a,
+                    b = get_parameters(object)$b,
+                    c = get_parameters(object)$c)
+
+    #function for RHS of ode system
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)),
+        {
+          S <- state[1:age]
+          E <- state[(age+1):(2*age)]
+          I <- state[(2*age+1):(3*age)]
+          R <- state[(3*age+1):(4*age)]
+          # rate of change
+          dS <- -a*S*I
+          dE <- a*S*I - b*E
+          dI <- b*E - c*I
+          dR <- c*I
+          # return the rate of change
+          list(c(dS, dE, dI, dR))
+        })
+    }
+
+    #call ode solver
+    out <- ode(
+      y = state, times = times, func = right_hand_side, parms = parameters)
+
+    output <- as.data.frame.array(out)
+
+    if(is_plot == TRUE){
+      col <- c('cS' = 'blue', 'cE' = 'green', 'cI' = 'yellow', 'cR' = 'red')
+      for(a in 1:age){
+        out_df <- output[,c(1, a+1, age+a+1, 2*age+a+1, 3*age+a+1)]
+        names(out_df) <- c('time', 'S', 'E', 'I', 'R')
+        SEIRplot <- ggplot(out_df, aes(x=time)) +
+          geom_line(aes(y=S, color = 'cS'), size = 1.5) +
+          geom_line(aes(y=E, color = 'cE'), size = 1.5) +
+          geom_line(aes(y=I, color = 'cI'), size = 1.5) +
+          geom_line(aes(y=R, color = 'cR'), size = 1.5) +
+          labs(x = 'time', y = 'fraction of the population',
+               title = glue('SEIR model for age group {a}')) +
+          theme(legend.position = 'right') +
+          scale_color_manual(values = col)
+        #show plot
+        print(SEIRplot)
+
+        #printtype of the plotting object to the command line for testing
+        print(typeof(SEIRplot))
+      }
+    }
+
+    # compute incidence number
+    total_inf <- output[, (2*age+2):(3*age+1)] + output[, (3*age+2):(4*age+1)]
```

Comment:
I'm not sure I understand what this code is doing here. I'm also a little wary of these hard-coded numeric indices: is there a way to make them strings so as to a) make it more readable and b) avoid issues with these hard coded values?

Reply:
See my comment below about returning an output in a different form, which would probably make it easier to avoid this type of hard-coded values issue.